### PR TITLE
Adds runge-kutta implementation

### PR DESCRIPTION
**Pull request by inivekin**
Added some of the fixes suggested from [here](https://github.com/inivekin/runge-kutta-4/issues/1#issue-830933371) as well as some logic fixes.

Interesting note: replacing ``` [ ] [ compose ] reduce ``` in ```(runge-kutta-4)``` with ``` concat ``` causes the ```lorenz.``` example to slow from ~0.13 to ~1.2 seconds. Haven't looked into that yet but I expect someone else will be able to figure out why faster than I can?

I think there are some repeated operations occurring within the follow loop that could be moved outside of it instead but I'm not certain it wouldn't hamper adapting it to other runge-kutta approximations.

### CODE REVIEW

File: extra/math/runge-kutta/runge-kutta.factor [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        ! using concat instead causes slow down by an order of magnitude
+        [ ] [ compose ] reduce
+
+        ! this will always produce 4 outputs with a duplication of the last result
+        ! NOTE the dup is kept in the transform so that function
+        ! can be used in higher-order estimation
+        call( -- rk1 rk2 rk3 rk4 rk4 ) drop 4array 
+
+        ! make array of zeroes of appropriate length to reduce into
+        dup first length>> 0 <repetition> >array
```
Comment (alex-ilin) [inline extra/math/runge-kutta/runge-kutta.factor:51]: It seems like `<repetiton>` is unnecessary here: `dup first length>> 0 <array>` should be enough.

File: extra/math/runge-kutta/runge-kutta.factor [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+: (runge-kutta) ( delta tx..n dx..n/dt -- rk )
+    [ swap ] dip dup length>> [ cleave ] dip narray swap v*n
+    ; inline
+
+: runge-kutta-differentials ( dx..n/dt -- seq )
+    '[ _ (runge-kutta) ] ;
+
+: runge-kutta-transforms ( tx..n delta dx..n/dt -- seq )
+    -rot swap
```
Comment (alex-ilin) [inline extra/math/runge-kutta/runge-kutta.factor:22]: Also known as `spin`.

File: extra/math/runge-kutta/runge-kutta.factor [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+IN: math.runge-kutta
+
+: runge-kutta-2-transform ( rk1 tx..n delta -- tx..n' delta )
+    [ swap [ [ v*n ] keep prefix 1/2 v*n ] dip v+ 2 v*n ] keep ;
+
+: runge-kutta-3-transform ( rk2 tx..n delta -- tx..n' delta )
+    runge-kutta-2-transform ;
+
+: runge-kutta-4-transform ( rk3 tx..n delta -- tx..n' delta )
+    [ [ swap ] dip [ v*n ] keep prefix v+ ] keep ;
```
Comment (alex-ilin) [inline extra/math/runge-kutta/runge-kutta.factor:12]: `[ swap ] dip` is `swapd`.

File: extra/math/runge-kutta/examples/examples.factor [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    tx..n rest first3 :> ( x y z )
+    y z 1 - x sq + * gamma x * + ;
+
+:: rf-dy/dt ( tx..n gamma -- dy )
+    tx..n rest first3 :> ( x y z )
+    x 3 z * 1 + x sq - * gamma y * + ;
+
+:: rf-dz/dt ( tx..n alpha -- dz )
+    tx..n rest first3 :> ( x y z )
+    2 neg z * alpha x y * + * ;
```
Comment (alex-ilin) [inline extra/math/runge-kutta/examples/examples.factor:39]: `2 neg`? Why not simply `-2`?

File: extra/math/runge-kutta/examples/examples.factor [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+       ui.gadgets ui.gadgets.charts ui.gadgets.panes ui.gadgets.charts.lines ui.theme ;
+IN: math.runge-kutta.examples
+
+: lorenz-dx/dt ( tx..n -- dx )
+    rest first2
+    swap - 10 * ;
+
+: lorenz-dy/dt ( tx..n -- dy )
+    rest first3
+    28 swap - [ swap ] dip * swap - ;
```
Comment (alex-ilin) [inline extra/math/runge-kutta/examples/examples.factor:11]: Another `swapd` here.

**Comment by mrjbq7** [comment]
Awesome!!

@inivekin let me know if you want me to wait to merge for any changes or if you're ready!

Thanks

**Comment by inivekin** [comment]
Made the changes suggested by @alex-ilin, thanks!

@mrjbq7 I don't intend to make any other changes for this pull request unless prompted to, feel free to merge when you want.

**Comment by mrjbq7** [comment]
Merged, thanks!!

**Pull request merged**